lib/aws/rails/middleware/elastic_beanstalk_sqsd.rb

```diff
@@ -28,25 +30,64 @@ def call(env)
           periodic_task?(request) ? execute_periodic_task(request) : execute_job(request)
         end
 
+        def shutdown(timeout = nil)
+          return unless @executor
+
+          @logger.info("Shutting down SQS EBS background job executor. Timeout: #{timeout}")
+          @executor.shutdown
+          clean_shutdown = @executor.wait_for_termination(timeout)
+          @logger.info("SQS EBS background executor shutdown complete. Clean: #{clean_shutdown}")
+        end
+
         private
 
+        def init_executor
+          threads = Integer(ENV.fetch('AWS_PROCESS_BEANSTALK_WORKER_THREADS',
+                                      Concurrent.available_processor_count || Concurrent.processor_count))
+          puts "Running with threads: #{threads}"
+          options = {
+            max_threads: threads,
+            max_queue: 1,
+            fallback_policy: :abort # Concurrent::RejectedExecutionError must be handled
+          }
+          @executor = Concurrent::ThreadPoolExecutor.new(options)
+          at_exit { shutdown }
+        end
+
         def execute_job(request)
+          if @executor
+            _execute_job_parallel(request)
+          else
+            _execute_job_now(request)
+          end
+        end
+
+        # Execute a job in the current thread
+        def _execute_job_now(request)
           # Jobs queued from the SQS adapter contain the JSON message in the request body.
           job = ::ActiveSupport::JSON.decode(request.body.string)
           job_name = job['job_class']
           @logger.debug("Executing job: #{job_name}")
-          _execute_job(job, job_name)
-          [200, { 'Content-Type' => 'text/plain' }, ["Successfully ran job #{job_name}."]]
-        rescue NameError
-          internal_error_response
-        end
-
-        def _execute_job(job, job_name)
           ::ActiveJob::Base.execute(job)
+          [200, { 'Content-Type' => 'text/plain' }, ["Successfully ran job #{job_name}."]]
         rescue NameError => e
           @logger.error("Job #{job_name} could not resolve to a class that inherits from Active Job.")
           @logger.error("Error: #{e}")
-          raise e
+          internal_error_response
+        end
+
+        # Execute a job using the thread pool executor
+        def _execute_job_parallel(request)
+          job_data = ::ActiveSupport::JSON.decode(request.body.string)
+          @executor.post(job_data) do |job|
+            @logger.debug("Executing job [in thread]: #{job['job_class']}")
+            ::ActiveJob::Base.execute(job)
+          end
+          [200, { 'Content-Type' => 'text/plain' }, ["Successfully queued job #{job_data['job_class']}"]]
```

Comment:
Should you also logger.info the messages in all of the success cases?

Reply:
activejob already performs its own logging of these (both foreground and background).  Example:
```
[ActiveJob] [TestJob] [252646fa-067a-40f3-a1b4-00f40d604eb0] Performing TestJob (Job ID: 252646fa-067a-40f3-a1b4-00f40d604eb0) from Sqs(default)

```